File: image/fabric/health.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0
"""Transport the hosted runtime's Fabric health report without invoking an agent.

Consumer contract: NeMo-Fabric PR #305 at d3aebd464458dcddeaf2965249154dbc4353c93e.
2026-09-16: bridge the proposed API; older SDKs explicitly remain unsupported.
"""

import asyncio
import json


def unavailable(reason):
    return {"supported": True, "report": None, "reason_code": reason}


async def runtime_health(runtime, timeout_seconds=3.0):
    if runtime is None:
        return unavailable("runtime_unavailable")
    check = getattr(runtime, "check_health", None)
    if check is None:
        return {"supported": False, "report": None, "reason_code": "fabric_health_unsupported"}
    try:
        result = await asyncio.wait_for(check(timeout_seconds=timeout_seconds), timeout_seconds)
        report = result.to_mapping()
        if report["runtime_id"] != runtime.runtime_id:
            raise ValueError("health runtime identity changed")
        return {"supported": True, "report": report, "reason_code": None}
    except TimeoutError:
        return unavailable("fabric_health_timeout")
    except Exception:
        return unavailable("fabric_health_error")


async def request_health(socket):
    async def request():
        reader, writer = await asyncio.open_unix_connection(socket, limit=65536)
        try:
            writer.write(b'{"operation":"health"}\n')
            await writer.drain()
            return json.loads(await reader.readline())
        finally:
            writer.close()
            await writer.wait_closed()

    try:
        return await asyncio.wait_for(request(), 5)
    except TimeoutError:
        return unavailable("health_transport_timeout")
    except Exception:
        return unavailable("health_transport_error")

File: crates/nemoclaw-sdk/src/health.rs
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

//! Consumer bridge for Fabric PR #305, revision d3aebd464458dcddeaf2965249154dbc4353c93e.
//! 2026-09-16: retain upstream reports without implementing adapter health semantics.
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RuntimeHealth {
    pub supported: bool,
    pub report: Option<Value>,
    pub reason_code: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct SandboxHealth {
    pub sandbox: String,
    pub agents: Vec<String>,
    #[serde(flatten)]
    pub health: RuntimeHealth,
}

impl RuntimeHealth {
    pub(crate) fn decode(bytes: &[u8]) -> Result<Self, crate::Error> {
        let health: Self = serde_json::from_slice(bytes).map_err(|_| {
            crate::Error::Conflict("invalid Fabric health response; resources retained")
        })?;
        if !health.valid() {
            return Err(crate::Error::Conflict(
                "invalid Fabric health response; resources retained",
            ));
        }
        Ok(health)
    }

    fn valid(&self) -> bool {
        let Some(report) = &self.report else {
            return match (self.supported, self.reason_code.as_deref()) {
                (false, Some("fabric_health_unsupported")) => true,
                (true, Some(reason)) => [
                    "runtime_unavailable",
                    "runtime_changed",
                    "fabric_health_timeout",
                    "fabric_health_error",
                    "health_transport_timeout",
                    "health_transport_error",
                ]
                .contains(&reason),
                _ => false,
            };
        };
        self.supported
            && self.reason_code.is_none()
            && report["runtime_id"].as_str().is_some_and(|s| !s.is_empty())
            && report["reason_code"]
                .as_str()
                .is_some_and(|s| !s.is_empty())
            && report["checked_at_millis"].as_u64().is_some()
            && report["duration_millis"].as_u64().is_some()
            && ["responsive", "unresponsive", "exited", "unknown"]
                .contains(&report["liveness"].as_str().unwrap_or(""))
            && ["idle", "busy", "stopping", "unknown"]
                .contains(&report["activity"].as_str().unwrap_or(""))
            && ["ready", "not_ready", "unknown"]
                .contains(&report["readiness"].as_str().unwrap_or(""))
            && report["checks"].as_array().is_some_and(|checks| {
                checks.iter().all(|check| {
                    check["name"].as_str().is_some_and(|s| !s.is_empty())
                        && check["reason_code"].as_str().is_some_and(|s| !s.is_empty())
                        && check["observed_at_millis"].as_u64().is_some()
                        && check["age_millis"].as_u64().is_some()
                        && ["ok", "failed", "unknown", "unsupported"]
                            .contains(&check["status"].as_str().unwrap_or(""))
                })
            })
    }

    /// Older Fabric retains legacy readiness checks, with health explicitly unsupported.
    pub fn allows_apply_completion(&self) -> bool {
        if !self.valid() {
            return false;
        }
        if !self.supported {
            return true;
        }
        self.report.as_ref().is_some_and(|report| {
            report["liveness"] == "responsive" && report["readiness"] == "ready"
        })
    }
}
